File: dev/hub/Raspi/dmcServer/src/provider/deviceFactory.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
// Project: DMC Server
// Date:	2015/Feb/11
// Author:	Carmelo J. Fdez-Agüera Tortosa
//
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#include "deviceFactory.h"
#include <core/comm/json/json.h>
#include <device/hue/hueLight.h>
#include <service/user/device/kodi/kodi.h>
#include <service/user/device/scene.h>
#include <iostream>

namespace dmc {

	namespace {
		// A wrapper for simplicity
		template<class JDevice_>
		struct JsonConstrucible {
			static std::pair<std::string, std::function<Device*(unsigned, const Json&)>> factory(const std::string& _label) {
				return std::make_pair(_label, [](unsigned _id, const Json& _data) -> Device* {
					return new JDevice_(_id, _data);
				});
			}
		};
	}

	//------------------------------------------------------------------------------------------------------------------
	DeviceFactory::DeviceFactory() {
		// Hue lights
		mFactories.insert(std::make_pair("HueLight", [](unsigned _id, const Json& _data) -> Device* {
			return new hue::Light(_id, _data["name"].asText(), _data["data"]["id"].asText());
		}));
		// Json based devices
		mFactories.insert(JsonConstrucible<Scene>::factory("Scene"));
		mFactories.insert(JsonConstrucible<kodi::Kodi>::factory("Kodi"));
	}

	//------------------------------------------------------------------------------------------------------------------
	Device* DeviceFactory::create(unsigned _id, const std::string& _devType, const Json& _data) {
		auto factory = mFactories.find(_devType);
		if(factory == mFactories.end()) {
			std::cout << "Unknown device type " << _devType << ". Ignoring device.\n";
			return nullptr;
		}
		else
			return factory->second(_id, _data);
	}

}	// namespace dmc

File: ui/app/android/app/src/main/java/app/dmc/devices/DeviceFactory.java
package app.dmc.devices;

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//
//      Domocracy Android App
//         Author: Pablo R.S.
//         Date:    2015-FEB-16
//
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
//

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import app.dmc.devices.supported_devices.HueLight;
import app.dmc.devices.supported_devices.Kodi;

public class DeviceFactory {
    //-----------------------------------------------------------------------------------------------------------------
    // Static Interface
    static public DeviceFactory get() {
        if(sDevFactory == null)
            sDevFactory = new DeviceFactory();

        return sDevFactory;
    }

    //-----------------------------------------------------------------------------------------------------------------
    //  Public Interface
    public Device create(String _type, JSONObject _data){
        return sFactories.get(_type).create(_data);
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Private Interface
    private DeviceFactory() {
        loadFactories();
    }

    private void loadFactories() {
        // HueLight Factory
        sFactories.put("HueLight", new Factory() {
            @Override
            public Device create(JSONObject _data) {
                return new HueLight(_data);
            }
        });

        // Scene Factory
        sFactories.put("Scene", new Factory() {
            @Override
            public Device create(JSONObject _data) {
                return new Scene(_data);
            }
        });

        // Kodi Factory
        sFactories.put("Kodi", new Factory() {
            @Override
            public Device create(JSONObject _data) {
                return new Kodi(_data);
            }
        });

        // Add other Factories
    }


    //-----------------------------------------------------------------------------------------------------------------
    // Members
    static private DeviceFactory sDevFactory = null;
    static private Map<String, Factory> sFactories =  new HashMap<>();

    //-----------------------------------------------------------------------------------------------------------------
    public interface Factory{
        Device create(JSONObject _data);
    }
}
